packages/core/src/components/button/abstractButton.tsx

```diff
@@ -40,31 +50,60 @@ export abstract class AbstractButton<T> extends React.Component<React.HTMLProps<
 
     public abstract render(): JSX.Element;
 
-    protected onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
-        switch (e.which) {
-            case Keys.SPACE:
-            case Keys.ENTER:
-                e.preventDefault();
-                if (e.which !== this.currentKeyDown) {
-                    this.setState({ isActive: true });
-                }
-                break;
-            default:
-                break;
+    protected getProps() {
+        const disabled = this.props.disabled || this.props.loading;
+
+        const className = classNames(
+            Classes.BUTTON, {
+                [Classes.ACTIVE]: this.state.isActive,
+                [Classes.DISABLED]: this.props.disabled || this.props.loading,
+                [Classes.LOADING]: this.props.loading,
+            },
+            Classes.iconClass(this.props.iconName),
+            Classes.intentClass(this.props.intent),
+            this.props.className,
+        );
+
+        return {
+            className,
+            disabled,
+            onClick: disabled ? undefined : this.props.onClick,
+            onKeyDown: this.handleKeyDown,
+            onKeyUp: this.handleKeyUp,
+            ref: this.refHandlers.button,
+        };
+    }
+
+    protected handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
+        if (isKeyboardClick(e.which)) {
+            e.preventDefault();
+            if (e.which !== this.currentKeyDown) {
+                this.setState({ isActive: true });
+            }
         }
         this.currentKeyDown = e.which;
     }
 
-    protected onKeyUp = (e: React.KeyboardEvent<HTMLElement>) => {
-        switch (e.which) {
-            case Keys.SPACE:
-            case Keys.ENTER:
-                this.setState({ isActive: false });
-                this.buttonRef.click();
-                break;
-            default:
-                break;
+    protected handleKeyUp = (e: React.KeyboardEvent<HTMLElement>) => {
+        if (isKeyboardClick(e.which)) {
+            this.setState({ isActive: false });
+            this.buttonRef.click();
         }
         this.currentKeyDown = null;
     }
+
+    protected renderChildren(): React.ReactNode {
+        const { children, loading, rightIconName, text } = this.props;
+        const iconClasses = classNames(Classes.ICON_STANDARD, Classes.iconClass(rightIconName), Classes.ALIGN_RIGHT);
+        return [
+            loading ? <Spinner className="pt-small pt-button-spinner" key="spinner" /> : undefined,
+            text != null ? <span key="text">{text}</span> : undefined,
+            children,
+            rightIconName != null ? <span className={iconClasses} key="icon" /> : undefined,
```

Comment:
I tend to agree with @cmslewis here, the old code looked more readable, but won't block on it